numpy/random/_examples/cython/extending_distributions.pyx

```diff
@@ -72,3 +74,43 @@ def uint10_uniforms(Py_ssize_t n):
     randoms = np.asarray(random_values)
     return randoms
 
+# cython example 3
+def uniforms_ex(bit_generator, Py_ssize_t n, dtype=np.float64):
+    """
+    Create an array of `n` uniformly distributed doubles via a "fill" function.
+
+    A 'real' distribution would want to process the values into
+    some non-uniform distribution
+
+    Parameters
+    ----------
+    bit_generator: BitGenerator instance
+    n: int
+        Output vector length
+    dtype: {str, dtype}, optional
+        Desired dtype, either 'd' (or 'float64') or 'f' (or 'float32'). The
+        default dtype value is 'd'
+    """
+    cdef Py_ssize_t i
+    cdef bitgen_t *rng
+    cdef const char *capsule_name = "BitGenerator"
+    cdef np.ndarray randoms
+
+    _dtype = np.dtype(dtype)
+    if _dtype.type not in (np.float32, np.float64):
+        raise TypeError('Unsupported dtype "%r"' % dtype)
+    capsule = bit_generator.capsule
+    # Optional check that the capsule if from a BitGenerator
+    if not PyCapsule_IsValid(capsule, capsule_name):
+        raise ValueError("Invalid pointer to anon_func_state")
+    # Cast the pointer
+    rng = <bitgen_t *> PyCapsule_GetPointer(capsule, capsule_name)
+    randoms = np.empty(n, dtype=_dtype)
+    if _dtype is np.float32:
```

Comment:
```suggestion
    if _dtype.type is np.float32:
```
or
```suggestion
    if _dtype == np.float32:
```
As written this is always false.

Reply:
Got be careful here, the latter is correct. The former would have to also check `_dtype.isnative`. Right now `dtype=">f8"` (on little endian machines) is probably broken!

Reply:
However, I think I actually like the `dtype.type is np.float32 and dtype.isnative` spelling I think.

Reply:
This is meant as a test-run for something to replace the `key = np.dtype(dtype).name` in `_generator.pyx` and `mtrand.pyx` so it would be good to get it right here.

Reply:
I think `dtype == np.float32` is probably simplest, but make sure to update line 100 too.

Reply:
True `dtype == ...` is shortest. May be worth to time it in case the type + isnative happens to be much faster (right now).
I timed it, and it:
```
In [2]: %timeit dt == np.float64                                                                                        
85 ns ± 5.03 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [3]: %timeit dt.type is np.float64                                                                                   
57.3 ns ± 0.108 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [4]: %timeit dt.type is np.float64 and dt.isnative                                                                   
81.6 ns ± 0.697 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
so since the `==` includes the isnative check there is no speed difference between the spellings , so probably should use the `==` one. The only thing we could consider with respect to fixing the speed issue of `dtype.name` check would be to add a fast-path for the very common dtype.

Reply:
using the `==` form

Reply:
Still seems to be the `==` form  here?

Reply:
fixing